### PR TITLE
Use the application ID we have already received

### LINKED ISSUE #139: Need to work around CF v2 endpoint bug when using v3 endpoints to deploy
## Environment
cf api version: 2.142

## Summary / Problem Statement
When an app is deployed using the 'rolling' deployment strategy Promregator can no long call it. This is due to a [bug in CF (cloud_controller_ng#1523)](https://github.com/cloudfoundry/cloud_controller_ng/issues/1523) making the /v2/space endpoint return the app's process GUID instead of the app's own GUID.

## Observed Behavior
Promregator reports 404s when calling the Prometheus endpoint for the give app instance

## Expected Behavior
It should be able to continue collecting metrics.

I'll begin working on a local workaround and submit a PR shortly (since we are busted at this point!). Since the v2 endpoints are being sunsetted, is there any concerns if I migrate directly to the v3 endpoints?

### PR DESCRIPTION
**Pull request by checketts**
Instead of the one from the space summary since the v2 endpoint
will sometimes be incorrect. Fixes #139

## Current Situation
The application ID is retrieved initially when resolving targets, then discarded. Later we store the application ID that is returned by the Space Summary call.

## Problem Statement
Due to a bug with the v2 Space Summary endpoint, the application ID we store will be incorrect, if the app was deployed using a 'rolling' strategy (which is a v3 endpoint mechanism)

## Solution Approach
I considered moving to the v3 endpoint, but the semantics are quite different. Instead I just cached the application ID we originally get.

### CODE REVIEW

File: src/main/java/org/cloudfoundry/promregator/scanner/ReactiveAppInstanceScanner.java [hunk 1]
```diff
@@ -182,9 +183,11 @@ public void setNumberOfInstances(int numberOfInstances) {
 				// NB: This drops the current target!
 				return Mono.empty();
 			}
-			
-			v.setApplicationId(sas.getId());
-			
+
+			if(v.getApplicationId() == null) {
```
Comment (eaglerainbow) [inline src/main/java/org/cloudfoundry/promregator/scanner/ReactiveAppInstanceScanner.java]: I am unsure whether this is necessary at all. If the `target` had not have an id in first place, then it would have been wrong by the `TargetResolver` to pass it through at all.  Hence, my impression is that this coding here is a little paranoid. 
Do you see a reason why it is necessary?

**Comment by eaglerainbow** [comment]
For the record: The PR uses the app id from the `TargetResolver` instead of taking the data from the space's summary information (what we were doing before). The `TargetResolver` retrieves it by calling

https://github.com/promregator/promregator/blob/df99cb90932876c59e01921a4b9997ae57069f3c/src/main/java/org/cloudfoundry/promregator/scanner/ReactiveTargetResolver.java#L347

or

https://github.com/promregator/promregator/blob/df99cb90932876c59e01921a4b9997ae57069f3c/src/main/java/org/cloudfoundry/promregator/scanner/ReactiveTargetResolver.java#L372

These two requests are supplied by the `appsInSpace` cache. Before, (using the space's summary cache), we were using the cache `spaceSummary` instead. *Fortunately*, the two caches share the same cache parameter:

https://github.com/promregator/promregator/blob/df99cb90932876c59e01921a4b9997ae57069f3c/src/main/java/org/cloudfoundry/promregator/cfaccessor/CFAccessorCacheClassic.java#L67

and

https://github.com/promregator/promregator/blob/df99cb90932876c59e01921a4b9997ae57069f3c/src/main/java/org/cloudfoundry/promregator/cfaccessor/CFAccessorCacheClassic.java#L68

So, we don't have to fear that the behavior would be perceived as incompatible by users, which have very strict cache timings configured (and now would have received to see a different cache behavior, e.g. after preforming a blue/green deployment with route-shift-over and two separate apps).

**Comment by eaglerainbow** [comment]
We should consider to downport this PR to `rel-0.7` and release a new patch version for it.

File: src/main/java/org/cloudfoundry/promregator/scanner/ReactiveAppInstanceScanner.java (hunk 1 above)
Comment (checketts) [reply src/main/java/org/cloudfoundry/promregator/scanner/ReactiveAppInstanceScanner.java]: Nope. It isn't necessary anymore. I'll delete it. Since this was my first contribution to the project I wasn't sure how paranoid you wanted me to be.

File: src/main/java/org/cloudfoundry/promregator/scanner/ResolvedTarget.java [hunk 2]
```diff
@@ -173,4 +175,12 @@ public String toString() {
 		builder.append("]");
 		return builder.toString();
 	}
```
[the remark this answers is not in the document]
Comment (checketts) [reply src/main/java/org/cloudfoundry/promregator/scanner/ResolvedTarget.java]: I'm using IntelliJ, so I'll use it's generator and check if it is similar, or hand code it if needed.

File: src/main/java/org/cloudfoundry/promregator/scanner/ReactiveAppInstanceScanner.java (hunk 1 above)
Comment (eaglerainbow) [reply src/main/java/org/cloudfoundry/promregator/scanner/ReactiveAppInstanceScanner.java]: Appreciated. 😁  
Paranoia isn't the worst habit for a developer, but that doesn't mean that we cannot rule things out...

File: src/main/java/org/cloudfoundry/promregator/scanner/ResolvedTarget.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (eaglerainbow) [reply src/main/java/org/cloudfoundry/promregator/scanner/ResolvedTarget.java]: Don't invest too much time into it.
Alternatively, I can generate them here just with a wave of the hand.

File: src/main/java/org/cloudfoundry/promregator/scanner/ReactiveAppInstanceScanner.java [hunk 3]
```diff
@@ -182,11 +183,14 @@ public void setNumberOfInstances(int numberOfInstances) {
 				// NB: This drops the current target!
 				return Mono.empty();
 			}
-			
-			v.setApplicationId(sas.getId());
-			
+
 			List<String> urls = sas.getUrls();
 			if (urls != null && !urls.isEmpty()) {
+				/*
```
Comment (eaglerainbow) [inline src/main/java/org/cloudfoundry/promregator/scanner/ReactiveAppInstanceScanner.java]: This comment looks misplaced to me.

File: src/main/java/org/cloudfoundry/promregator/scanner/ReactiveAppInstanceScanner.java [hunk 4]
```diff
@@ -182,9 +183,11 @@ public void setNumberOfInstances(int numberOfInstances) {
 				// NB: This drops the current target!
 				return Mono.empty();
 			}
-			
-			v.setApplicationId(sas.getId());
-			
+
```
[the remark this answers is not in the document]
Comment (checketts) [reply src/main/java/org/cloudfoundry/promregator/scanner/ReactiveAppInstanceScanner.java]: Since I remove the id assignment. I wasn't certain where to place this warning. I placed it a few lines down where we fetch the urls from the SAS. I'm fine removing it if you like.

File: src/main/java/org/cloudfoundry/promregator/scanner/ReactiveAppInstanceScanner.java (hunk 3 above)
Comment (checketts) [reply src/main/java/org/cloudfoundry/promregator/scanner/ReactiveAppInstanceScanner.java]: Since the ID assignment is removed. I'm not certain where to place this either. I could put it at the point where we retrieve the SAS. Or just remove it.

Comment (eaglerainbow) [reply src/main/java/org/cloudfoundry/promregator/scanner/ReactiveAppInstanceScanner.java]: I'd put it [after line 186](https://github.com/promregator/promregator/pull/140/files#diff-eac88078a44f3aaf65fdd4858abe390bR186), so right before the first usage of any attribute, but after it's clear that it has been initialized properly. 

What do you think?

**Comment by eaglerainbow** [comment]
Looks good to me now. 
I'll trigger the voter manually now. Note that it'll need ~28 mins to complete.

**Comment by checketts** [comment]
I cherry picked the changes against rel-0.7. Should i just create a new PR against that branch?

**Comment by eaglerainbow** [comment]
>I cherry picked the changes against rel-0.7. Should i just create a new PR against that branch?

Wait until I have merged this PR.
I don't want to create you unnecessary effort.

**Comment by eaglerainbow** [comment]
The voter crashed strangely. Lemme check...

**Comment by eaglerainbow** [comment]
Mhm... that's not good: This enhancement has implications to the unit tests (assertions do not hold true anymore).

```
java.lang.AssertionError
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertTrue(Assert.java:52)
	at org.cloudfoundry.promregator.discovery.CFMultiDiscovererTest.testDiscoverWithCleanup(CFMultiDiscovererTest.java:73)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.springframework.test.context.junit4.statements.RunBeforeTestExecutionCallbacks.evaluate(RunBeforeTestExecutionCallbacks.java:74)
	at org.springframework.test.context.junit4.statements.RunAfterTestExecutionCallbacks.evaluate(RunAfterTestExecutionCallbacks.java:84)
	at org.springframework.test.context.junit4.statements.RunBeforeTestMethodCallbacks.evaluate(RunBeforeTestMethodCallbacks.java:75)
	at org.springframework.test.context.junit4.statements.RunAfterTestMethodCallbacks.evaluate(RunAfterTestMethodCallbacks.java:86)
	at org.springframework.test.context.junit4.statements.SpringRepeat.evaluate(SpringRepeat.java:84)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:251)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:97)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:70)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:190)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:86)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:538)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:760)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:460)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:206)
```

```
java.lang.NullPointerException
	at org.cloudfoundry.promregator.scanner.MassReactiveAppInstanceScannerTest.testPerformanceWithApplicationIdAndInstanceFilter(MassReactiveAppInstanceScannerTest.java:145)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.springframework.test.context.junit4.statements.RunBeforeTestExecutionCallbacks.evaluate(RunBeforeTestExecutionCallbacks.java:74)
	at org.springframework.test.context.junit4.statements.RunAfterTestExecutionCallbacks.evaluate(RunAfterTestExecutionCallbacks.java:84)
	at org.springframework.test.context.junit4.statements.RunBeforeTestMethodCallbacks.evaluate(RunBeforeTestMethodCallbacks.java:75)
	at org.springframework.test.context.junit4.statements.RunAfterTestMethodCallbacks.evaluate(RunAfterTestMethodCallbacks.java:86)
	at org.springframework.test.context.junit4.statements.SpringRepeat.evaluate(SpringRepeat.java:84)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:251)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:97)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:70)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:190)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:86)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:538)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:760)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:460)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:206)

```


plus all tests in `ReactiveAppInstanceScannerTest`

Do you want to check them or should I have a look?

**Comment by checketts** [comment]
I already had to fix `ReactiveAppInstanceScannerTest` when running locally. But I was a little stumped since they were failing on master. I'll push my local fixes and comment on my analysis in the PR.

**Comment by eaglerainbow** [comment]
>  But I was a little stumped since they were failing on master

Then let me check... this should not be the case.

**Comment by checketts** [comment]
Oh sorry it was the `CFAccessorCacheClassicTest` that had failed for me before. So this is new

**Comment by checketts** [comment]
OK I see the problem with `ReactiveAppInstanceScannerTest` this will be a quick fix

**Comment by eaglerainbow** [comment]
Unit tests on 3330cd742f3710b89a855cdf80c47f2e2f116f0c run fine for me (tested locally).

Note that f0eacbd69093c880384601ddf9d05ea8dda2b17a has parent 94688a4530a6d58d12e4adf74c21ab68c3b31a82

**Comment by checketts** [comment]
Sorry gonna take a second to rewrite these assertions with assertj instead just fixing the tests so we'll be able to be able to read them better

**Comment by checketts** [comment]
Nevermind. I forgot that java 8 stream have so much more ceremony than a kotlin flter. 😨

**Comment by eaglerainbow** [comment]
> Sorry gonna take a second to rewrite these assertions with assertj instead just fixing the tests so we'll be able to be able to read them better

No worries. This isn't a competition who's fixing it faster...

**Comment by eaglerainbow** [comment]
Voter is triggered

**Comment by eaglerainbow** [comment]
The voter says "two more red tests to go":

``` 
java.lang.AssertionError
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertTrue(Assert.java:52)
	at org.cloudfoundry.promregator.discovery.CFMultiDiscovererTest.testDiscoverWithCleanup(CFMultiDiscovererTest.java:73)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.springframework.test.context.junit4.statements.RunBeforeTestExecutionCallbacks.evaluate(RunBeforeTestExecutionCallbacks.java:74)
	at org.springframework.test.context.junit4.statements.RunAfterTestExecutionCallbacks.evaluate(RunAfterTestExecutionCallbacks.java:84)
	at org.springframework.test.context.junit4.statements.RunBeforeTestMethodCallbacks.evaluate(RunBeforeTestMethodCallbacks.java:75)
	at org.springframework.test.context.junit4.statements.RunAfterTestMethodCallbacks.evaluate(RunAfterTestMethodCallbacks.java:86)
	at org.springframework.test.context.junit4.statements.SpringRepeat.evaluate(SpringRepeat.java:84)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:251)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:97)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:70)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:190)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:86)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:538)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:760)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:460)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:206)


```

and 

```
java.lang.NullPointerException
	at org.cloudfoundry.promregator.scanner.MassReactiveAppInstanceScannerTest.testPerformanceWithApplicationIdAndInstanceFilter(MassReactiveAppInstanceScannerTest.java:145)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.springframework.test.context.junit4.statements.RunBeforeTestExecutionCallbacks.evaluate(RunBeforeTestExecutionCallbacks.java:74)
	at org.springframework.test.context.junit4.statements.RunAfterTestExecutionCallbacks.evaluate(RunAfterTestExecutionCallbacks.java:84)
	at org.springframework.test.context.junit4.statements.RunBeforeTestMethodCallbacks.evaluate(RunBeforeTestMethodCallbacks.java:75)
	at org.springframework.test.context.junit4.statements.RunAfterTestMethodCallbacks.evaluate(RunAfterTestMethodCallbacks.java:86)
	at org.springframework.test.context.junit4.statements.SpringRepeat.evaluate(SpringRepeat.java:84)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:251)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.runChild(SpringJUnit4ClassRunner.java:97)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.springframework.test.context.junit4.statements.RunBeforeTestClassCallbacks.evaluate(RunBeforeTestClassCallbacks.java:61)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.springframework.test.context.junit4.statements.RunAfterTestClassCallbacks.evaluate(RunAfterTestClassCallbacks.java:70)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.springframework.test.context.junit4.SpringJUnit4ClassRunner.run(SpringJUnit4ClassRunner.java:190)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:86)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:538)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:760)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:460)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:206)
```

**Comment by checketts** [comment]
The failure on the `CFMultiDiscovererTest` is due to using the `TrivialTargetRespolver` since it doesn't have the AppID in the original `Target`. I can solve that by adding that property to the `Target` or add the fallback code back in to looks up the AppID in the SpaceSummary (SAS). Which would you prefer?

The simplest would be keeping the fallback in place. However, it leaves the user exposed to the bug.

**Comment by eaglerainbow** [comment]
> The simplest would be keeping the fallback in place. However, it leaves the user exposed to the bug.

Yes, I am no fan of the fallback coding in production. 
Moreover, even if we had it in place, we would not win much, as still the id in the `ResolvedTarget` be `null`.

My thoughts are currently going into that direction to make the `TrivialTargetResolver ` not so trivial anymore: Your change expect that it does answer truthfully, which it does not do right now.

**Comment by eaglerainbow** [comment]
Last comment for tonight from my side:
The voter only complains about the case of `CFMultiDiscovererTest.testDiscoverWithCleanup:73`.

File: src/main/java/org/cloudfoundry/promregator/config/Target.java [hunk 5]
```diff
@@ -203,6 +215,8 @@ public String toString() {
 		builder.append(applicationName);
 		builder.append(", applicationRegex=");
 		builder.append(applicationRegex);
+		builder.append(", applicationId=");
+		builder.append(applicationId);
```
Comment (checketts) [inline src/main/java/org/cloudfoundry/promregator/config/Target.java]: I added 'applicationId' to the `Target` and leveraged it to fix the failing test. I didn't add the new property to `equals` or `hashCode` since they hadn't been overridden in `Target` yet.

Comment (eaglerainbow) [reply src/main/java/org/cloudfoundry/promregator/config/Target.java]: I think adding `applicationId` is dangerous: It permits users to explicitly set the applicationId in their promregator configuration file (NB: `org.cloudfoundry.promregator.config.Target` is used as POJO for configuration serialization/deserialization!)

File: src/main/java/org/cloudfoundry/promregator/scanner/ResolvedTarget.java [hunk 6]
```diff
@@ -14,6 +14,8 @@
 	private String path;
 	
 	private String protocol;
+
+	private String id;
```
Comment (eaglerainbow) [inline src/main/java/org/cloudfoundry/promregator/scanner/ResolvedTarget.java]: I just got here some second thoughts: `id` is very generic --- and it would not be the id of a `ResolvedTarget`. 
Can we agree on renaming it to `applicationId`?

Comment (checketts) [reply src/main/java/org/cloudfoundry/promregator/scanner/ResolvedTarget.java]: Happy to rename it. I'll do that first think when I get into work.

File: src/main/java/org/cloudfoundry/promregator/config/Target.java (hunk 5 above)
Comment (checketts) [reply src/main/java/org/cloudfoundry/promregator/config/Target.java]: It is dangerous. However it is the only was to have `TrivialTargetResolver` resolve the appId. I could fix the failing test by replacing `TrivialTargetResolver` with a mock that provides the correct IDs, but that indicated that `TrivialTargetResolver` doesn't work anymore and we probably should delete it.

Again, the other option of having the fallback to looking up the appId via the SpaceSummary would allow us to not add appId to Target, but that seemed like you wanted to avoid that.

File: src/main/java/org/cloudfoundry/promregator/scanner/ResolvedTarget.java (hunk 6 above)
Comment (checketts) [reply src/main/java/org/cloudfoundry/promregator/scanner/ResolvedTarget.java]: This change is done now.

File: src/main/java/org/cloudfoundry/promregator/config/Target.java (hunk 5 above)
Comment (checketts) [reply src/main/java/org/cloudfoundry/promregator/config/Target.java]: I haven't made a change here yet, since I'm not certain how you want me to proceed. I don't consider it a dangerous change for the users to be able to define the applicationId in their config, since it is ignored in the typical case and is now required if they are using the `TrivialTargetResolver` (I'm not certain how that resolver get activated, or if it is merely a testing helper)

If `TrivialTargetResolver` is only intended for testing, let me know. I will happily remove it and update the tests to no longer need it. That way I can remove the `applicationId` from the `Target` class.

**Comment by eaglerainbow** [comment]
@checketts Sorry, I am currently busy (today and tomorrow) with some other stuff and did not find time to have another look. Will do on next chance and will report back.

File: src/main/java/org/cloudfoundry/promregator/config/Target.java (hunk 5 above)
Comment (eaglerainbow) [reply src/main/java/org/cloudfoundry/promregator/config/Target.java]: > If TrivialTargetResolver is only intended for testing, let me know.

Given that `TrivialTargetResolver` is under `src/test/java` I consider it as "test-only" coding. 
I would go even one step further: If the tests can no longer be performed with a `TrivialTargetResolver` (to whatever reason), then this class has reached its end-of-life and shall be replaced with "something more suitable". 

>  I don't consider it a dangerous change for the users to be able to define the applicationId in their config

I see it as false sense of contract: In fact, this attribute is intended to be "internal only" - we don't really mean anyone to tell us the app's id we shall scrape --- and even if they did, we still would have to read the same set of metadata as before, as we require the app's name as well (e.g. for Enrichment).
In short, I don't like that we extend a public contract with an attribute that we don't mean like this. I clearly prefer the way of putting `TrivialTargetResolver` through the mill instead.

Agreed?

**Comment by eaglerainbow** [comment]
Could have a look at your question - see comment above...

File: src/main/java/org/cloudfoundry/promregator/config/Target.java (hunk 5 above)
Comment (checketts) [reply src/main/java/org/cloudfoundry/promregator/config/Target.java]: Wow. How did I overlook `TrivialTargetResolver` was a test only mechanism?  I'm happy to get rid of it and revert my changes to `Target`! I'll get right on that fix.

Comment (eaglerainbow) [reply src/main/java/org/cloudfoundry/promregator/config/Target.java]: thx a million!

File: src/test/java/org/cloudfoundry/promregator/discovery/CFMultiDiscovererTest.java [hunk 7]
```diff
@@ -45,6 +55,18 @@ public void receiver(Instance instance) {
...
+		ResolvedTarget aTarget = new ResolvedTarget();
+		aTarget.setOrgName("unittestorg");
+		aTarget.setSpaceName("unittestspace");
+		aTarget.setApplicationName("testapp");
+		aTarget.setApplicationId(CFAccessorMock.UNITTEST_APP1_UUID);
+		aTarget.setProtocol("https");
+		aTarget.setPath("/metrics");
+		aTarget.setOriginalTarget(new Target());
+		resolvedTargets.add(aTarget);
+		when(targetResolver.resolveTargets(any())).thenReturn(resolvedTargets);
```
Comment (checketts) [inline src/test/java/org/cloudfoundry/promregator/discovery/CFMultiDiscovererTest.java]: I replaced the `TrivialTargetResolver` with this mock and the tests are passing again

**Comment by eaglerainbow** [comment]
Waiting for the "go" by voter...
